syncd/syncd_mdio_ipc.cpp

```diff
@@ -0,0 +1,440 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <ctype.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/select.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <time.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <pthread.h>
+
+extern "C" {
+#include <sai.h>
+#include <saiversion.h>
+}
+
+#include "swss/logger.h"
+
+#define SYNCD_IPC_SOCK_SYNCD  "/var/run/sswsyncd"
+#define SYNCD_IPC_SOCK_HOST   "/var/run/docker-syncd"
+#define SYNCD_IPC_SOCK_FILE   "mdio-ipc"
+#define SYNCD_IPC_BUFF_SIZE   256   /* buffer size */
+
+#define CONN_TIMEOUT        30     /* sec, connection timeout */
+#define CONN_MAX            18     /* max. number of connections */
+
+#ifndef COUNTOF
+#define COUNTOF(x)          ((int)(sizeof((x)) / sizeof((x)[0])))
+#endif
+
+extern sai_object_id_t mdioSwitchId;
+
+static pthread_t taskId;
+static int taskAlive = 0;
+
+typedef struct syncd_mdio_ipc_conn_s
+{
+    int    fd;
+    time_t timeout;
+} syncd_mdio_ipc_conn_t;
+
+typedef struct syncd_mdio_ipc_cmd_s
+{
+    const char *key;
+    int (*func)(char *resp, int argc, char *argv[], sai_switch_api_t *sai_switch_api);
+} syncd_mdio_ipc_cmd_t;
+
+/* Remove leading and trailing spaces */
+static char *strim(char *s)
+{
+    size_t len;
+    char *end;
+
+    /* Remove the leading spaces */
+    while (isspace(*s))
+    {
+        ++s;
+    }
+
+    /* Remove the trailing spaces */
+    len = strlen(s);
+    if (len == 0)
+    {
+        return s;
+    }
+    end = s + len - 1;
+    while ((end >= s) && isspace(*end))
+    {
+        end--;
+    }
+    *(end + 1) = '\0';
+
+    return s;
+}
+
+/* Convert to lowercase string */
+static char *strlower(char *s)
+{
+    unsigned char *c;
+    unsigned char *str = (unsigned char *)s;
+
+    for (c = str; *c != 0; ++c)
+    {
+        *c = (unsigned char)tolower(*c);
+    }
+    return s;
+}
+
+/*
+ * mdio <port_oid> reg:     Read from the PHY register
+ * mdio <port_oid> reg val: Write to the PHY register
+ */
+static sai_status_t syncd_ipc_cmd_mdio_common(char *resp, int argc, char *argv[], sai_switch_api_t *sai_switch_api)
+{
+    int ret = 0;
+    uint32_t mdio_addr = 0, reg_addr = 0, val = 0;
+
+    if (argc < 3)
+    {
+        return SAI_STATUS_INVALID_PARAMETER;
+    }
+
+    mdio_addr = (uint32_t)strtoul(argv[1], NULL, 0);
+    reg_addr = (uint32_t)strtoul(argv[2], NULL, 0);
+
+    if (mdioSwitchId == SAI_NULL_OBJECT_ID)
+    {
+        SWSS_LOG_ERROR("mdioSwitchId not initialized");
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (argc > 3)
+    {
+        val = (uint32_t)strtoul(argv[3], NULL, 0);
+        ret = sai_switch_api->switch_mdio_write(mdioSwitchId, mdio_addr, reg_addr, 1, &val);
+        sprintf(resp, "%d\n", ret);
+    }
+    else
+    {
+        ret = sai_switch_api->switch_mdio_read(mdioSwitchId, mdio_addr, reg_addr, 1, &val);
+        sprintf(resp, "%d 0x%x\n", ret, val);
```

Comment:
output to a unknown buffer size out the function scope, use std::string to produce response or log errors here instead of passing it through the stack